File: engines/popular-recommend-engine/src/main/scala/io.sugo.pio.engine.popular/PopularEngineFactory.scala
package io.sugo.pio.engine.popular.engine

import io.sugo.pio.engine.common.data.QueryableModelData
import io.sugo.pio.engine.popular.data.{PopularModelData, PopularPreparaData, PopularTrainingData}
import io.sugo.pio.spark.engine._
import io.sugo.pio.spark.engine.data.input.{BatchEventHose, PropertyHose}

/**
  */
class PopularEngineFactory(propertyHose: PropertyHose, batchEventHose: BatchEventHose) extends EngineFactory[PopularTrainingData, PopularPreparaData, PopularModelData, QueryableModelData]{
  override def createDatasource(): DataSource[PopularTrainingData] = {
    val dataSource = new PopularDatasource(propertyHose, batchEventHose)
    dataSource
  }

  override def createPreparator(): Preparator[PopularTrainingData, PopularPreparaData] = {
    val preparator = new PopularPreparator()
    preparator
  }

  override def createAlgorithm(): Algorithm[PopularPreparaData, PopularModelData] = {
    val algorithm = new PopularAlgorithm()
    algorithm
  }

  override def createModel(): Model[PopularModelData, QueryableModelData] = {
    val model = new PopularModel
    model
  }
}

File: engines/popular-recommend-engine/src/main/java/io/sugo/pio/engine/popular/PopularEngineFactory.java
package io.sugo.pio.engine.popular;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import io.sugo.pio.engine.data.input.BatchEventHose;
import io.sugo.pio.engine.data.input.PropertyHose;
import io.sugo.pio.engine.data.output.Repository;
import io.sugo.pio.engine.popular.data.PopularModelData;
import io.sugo.pio.engine.popular.data.PopularPreparaData;
import io.sugo.pio.engine.popular.data.PopularTrainingData;
import io.sugo.pio.engine.training.*;

/**
 */
public class PopularEngineFactory implements EngineFactory {
    private final BatchEventHose batchEventHose;
    private final PropertyHose propertyHose;
    private final Repository repository;
    private final PopEngineParams engineParams;

    @JsonCreator
    public PopularEngineFactory(@JsonProperty("propertyHose") PropertyHose propertyHose,
                           @JsonProperty("batchEventHose") BatchEventHose batchEventHose,
                           @JsonProperty("repository") Repository repository,
                           @JsonProperty("engineParams") PopEngineParams engineParams) {
        this.batchEventHose = batchEventHose;
        this.propertyHose = propertyHose;
        this.repository = repository;
        this.engineParams = engineParams;
    }


    @Override
    public Engine createEngine() {
        return new PopularEngine(propertyHose, batchEventHose, repository, engineParams);
    }

    @JsonProperty
    public BatchEventHose getBatchEventHose() {
        return batchEventHose;
    }

    @JsonProperty
    public PropertyHose getPropertyHose() {
        return propertyHose;
    }

    @JsonProperty
    public PopEngineParams getEngineParams() {
        return engineParams;
    }

    @JsonProperty
    public Repository getRepository() {
        return repository;
    }
}
